<a href="https://colab.research.google.com/github/Adlucent/ga4-return-prediction/blob/main/5_Multicollinearity_Check_and_Joining_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 5. Multicollinearity Check and Join Data
At this point, we should check our features for multicollinearity issues.

In [ ]:
################################################################################
######################### CHANGE BQ PROJECT NAME BELOW #########################
################################################################################

project_name = 'adl-analytics' #add proj name
region = "US"  # GCP project region
customer_table_name = 'adl-analytics.return_prediction_ga4.step_4_customer_agg'
transaction_table_name = 'adl-analytics.return_prediction_ga4.step_4_split'

In [ ]:
# If your notebook does not have pandas_gbq you can install it here:
# ! pip install pandas_gbq

In [ ]:
# Google credentials
from google.colab import auth
auth.authenticate_user()

# # BigQuery Magics
# '''BigQuery magics are used to run BigQuery SQL queries in a python environment.
# These queries can also be run in the BigQuery UI '''

# from google.cloud import bigquery
# from google.cloud.bigquery import magics, Client, QueryJobConfig

# magics.context.project = project_name #update project name
# client = bigquery.Client(project=magics.context.project)

# Interface between Jupyter and BigQuery
import pandas_gbq

# data processing libraries
import pandas as pd
import numpy as np

# multicollinearity tools
from statsmodels.stats.outliers_influence import variance_inflation_factor

# visualization
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

# suppress notebook warnings
import warnings
warnings.filterwarnings('ignore')

# dataframe formatting
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.set_option("display.float_format", lambda x: "%.3f" % x)

# Load customer data

In [ ]:
sql = f"""
SELECT *
FROM `{customer_table_name}`;
"""
customer_data = pandas_gbq.read_gbq(sql, project_id=project_name, location=region, use_bqstorage_api=True)

Downloading: 100%|██████████|


In [ ]:
# create a copy of the imported data to avoid re-importing if we need to revert to original table
df_customer = customer_data.copy()

In [ ]:
df_customer.head()

user_pseudo_id ecommerce_transaction_id  \
0   1219636.7573278529                   775080   
1  12291446.3815334458                   205482   
2  12291446.3815334458                   597612   
3  13756133.7762491194                   462859   
4   1415699.2356865981                   484939   

   nunique_ecommerce_transactions  avg_event_params_engagement_time_msec  \
0                           2.000                               3754.163   
1                           1.000                               9148.439   
2                           2.000                              10078.185   
3                           2.000                               8643.410   
4                           1.000                               4372.630   

   avg_item_price_in_usd  avg_item_promotions  avg_item_quantity  \
0                  8.625                0.292              1.250   
1                 28.857                0.000              1.000   
2                 18.529                0.000              1.000   
3                 19.000                1.000              9.500   
4                 29.000                1.000              1.444   

   avg_item_refund_in_usd  avg_item_refund_quantity  avg_item_revenue_in_usd  \
0                  10.125                     1.250                   10.125   
1                  28.857                     1.000                   28.857   
2                  18.529                     1.000                   18.529   
3                  38.000                     2.000                  182.500   
4                   0.000                     0.000                   29.444   

   sum_ecommerce_total_item_quantity  sum_total_return_item_quantity  \
0                                  9                               9   
1                                  7                               7   
2                                 12                              12   
3                                 19                               4   
4                                 13                               0   

   sum_ecommerce_purchase_revenue_in_usd  sum_ecommerce_refund_value_in_usd  \
0                                 70.000                             70.000   
1                                203.000                            203.000   
2                                244.000                            244.000   
3                                365.000                             76.000   
4                                265.000                              0.000   

   sum_ecommerce_tax_value_in_usd  sum_ecommerce_unique_items  \
0                           9.000                           7   
1                          19.000                           7   
2                          24.000                          12   
3                          36.000                           2   
4                          23.000                           9   

   max_days_first_session_to_transaction  max_user_ltv_revenue  \
0                                  0.000                71.000   
1                                 10.000               824.000   
2                                 10.000               865.000   
3                                  4.000               365.000   
4                                  0.000               265.000   

   max_event_params_engagement_time_msec  max_item_price_in_usd  \
0                                  95696                 20.000   
1                                 186421                 44.000   
2                                 307177                 44.000   
3                                 332905                 19.000   
4                                  58902                 92.000   

   max_item_promotions  max_item_quantity  max_item_refund_in_usd  \
0                    1                  2                  20.000   
1                    0                  1                  44.000   
2                    0                  1       

In [ ]:
df_customer.shape

(4466, 226)

In [ ]:
# check for missing values
df_customer.isna().sum().sum()

0

# Customer VIF analysis

In [ ]:
# remove id and split columns from analysis
customer_features = df_customer.iloc[: , 2:-1]
customer_features.head()

nunique_ecommerce_transactions  avg_event_params_engagement_time_msec  \
0                           2.000                               3754.163   
1                           1.000                               9148.439   
2                           2.000                              10078.185   
3                           2.000                               8643.410   
4                           1.000                               4372.630   

   avg_item_price_in_usd  avg_item_promotions  avg_item_quantity  \
0                  8.625                0.292              1.250   
1                 28.857                0.000              1.000   
2                 18.529                0.000              1.000   
3                 19.000                1.000              9.500   
4                 29.000                1.000              1.444   

   avg_item_refund_in_usd  avg_item_refund_quantity  avg_item_revenue_in_usd  \
0                  10.125                     1.250                   10.125   
1                  28.857                     1.000                   28.857   
2                  18.529                     1.000                   18.529   
3                  38.000                     2.000                  182.500   
4                   0.000                     0.000                   29.444   

   sum_ecommerce_total_item_quantity  sum_total_return_item_quantity  \
0                                  9                               9   
1                                  7                               7   
2                                 12                              12   
3                                 19                               4   
4                                 13                               0   

   sum_ecommerce_purchase_revenue_in_usd  sum_ecommerce_refund_value_in_usd  \
0                                 70.000                             70.000   
1                                203.000                            203.000   
2                                244.000                            244.000   
3                                365.000                             76.000   
4                                265.000                              0.000   

   sum_ecommerce_tax_value_in_usd  sum_ecommerce_unique_items  \
0                           9.000                           7   
1                          19.000                           7   
2                          24.000                          12   
3                          36.000                           2   
4                          23.000                           9   

   max_days_first_session_to_transaction  max_user_ltv_revenue  \
0                                  0.000                71.000   
1                                 10.000               824.000   
2                                 10.000               865.000   
3                                  4.000               365.000   
4                                  0.000               265.000   

   max_event_params_engagement_time_msec  max_item_price_in_usd  \
0                                  95696                 20.000   
1                                 186421                 44.000   
2                                 307177                 44.000   
3                                 332905                 19.000   
4                                  58902                 92.000   

   max_item_promotions  max_item_quantity  max_item_refund_in_usd  \
0                    1                  2                  20.000   
1                    0                  1                  44.000   
2                    0                  1                  44.000   
3                    1                 10                  76.000   
4                    1                  3                   0.000   

   max_item_refund_quantity  max_item_revenue_in_usd  \
0                         2                   21.000   
1                         1   

In [ ]:
customer_features = customer_features.astype('float64')

In [ ]:
# first, let's take a look at VIF scores for all original features in the customer dataset
vif_before = pd.DataFrame(
    [variance_inflation_factor(customer_features.values, i) for i in range(customer_features.shape[1])],
    index=customer_features.columns,
    dtype=float,
).reset_index()
vif_before.columns = ('Feature', 'VIF')

In [ ]:
print('Before feature selection:')
vif_before

Before feature selection:


Feature        VIF
0                       nunique_ecommerce_transactions   5053.236
1                avg_event_params_engagement_time_msec     26.310
2                                avg_item_price_in_usd    103.844
3                                  avg_item_promotions     12.451
4                                    avg_item_quantity     62.263
5                               avg_item_refund_in_usd     32.489
6                             avg_item_refund_quantity     30.697
7                              avg_item_revenue_in_usd    118.084
8                    sum_ecommerce_total_item_quantity    579.435
9                       sum_total_return_item_quantity    223.094
10               sum_ecommerce_purchase_revenue_in_usd    317.376
11                   sum_ecommerce_refund_value_in_usd    278.124
12                      sum_ecommerce_tax_value_in_usd     69.061
13                          sum_ecommerce_unique_items    173.886
14               max_days_first_session_to_transaction      3.148
15                                max_user_ltv_revenue     32.103
16               max_event_params_engagement_time_msec      5.317
17                               max_item_price_in_usd    118.926
18                                 max_item_promotions     13.138
19                                   max_item_quantity     39.155
20                              max_item_refund_in_usd     58.880
21                            max_item_refund_quantity     50.085
22                             max_item_revenue_in_usd    155.101
23               sum_event_params_engagement_time_msec     86.435
24                    sum_event_params_session_engaged 144511.630
25                               sum_item_price_in_usd    581.340
26                                 sum_item_promotions     14.433
27                                   sum_item_quantity    172.792
28                              sum_item_refund_in_usd    269.033
29                            sum_item_refund_quantity    233.674
30                             sum_item_revenue_in_usd    886.009
31                          sum_event_name_add_to_cart   4819.337
32                       sum_event_name_begin_checkout   1666.267
33                            sum_event_name_page_view  18304.030
34                               sum_event_name_scroll    914.985
35                          sum_event_name_select_item   2119.156
36                      sum_event_name_user_engagement  11428.042
37                            sum_event_name_view_item  21972.691
38                         sum_device_category_desktop   2720.228
39                          sum_device_category_mobile    418.813
40                  sum_device_mobile_brand_name_Apple        inf
41                 sum_device_mobile_brand_name_Google    465.256
42                 sum_device_mobile_brand_name_Huawei      2.420
43              sum_device_mobile_brand_name_Microsoft        inf
44                sum_device_mobile_brand_name_Mozilla        inf
45                sum_device_mobile_brand_name_Samsung      4.852
46                 sum_device_mobile_brand_name_Xiaomi      7.222
47                 sum_device_mobile_model_name_Chrome    760.582
48             sum_device_mobile_model_name_ChromeBook     26.531
49                   sum_device_mobile_model_name_Edge        inf
50                sum_device_mobile_model_name_Firefox        inf
51                 sum_device_mobile_model_name_Safari        inf
52                   sum_device_mobile_model_name_iPad        inf
53                 sum_device_mobile_model_name_iPhone        inf
54          sum_device_web_info_browser_AndroidWebview     19.182
55                  sum_device_web_info_browser_Chrome    687.085
56                    sum_device_web_info_browser_Edge     17.979
57                 sum_device_web_info_browser_Firefox        inf
58                  sum_device_web_info_browser_Safari     99.341
59                              sum_geo_country_Canada     19.461
60                   

In [ ]:
vif_before[vif_before['VIF'] > 5].shape[0]

183

Since so many of our features have such high VIF values, let's start by dropping the features with a VIF value over 1000 and recalculate VIF.

In [ ]:
# drop 40 features with VIF over 1000
customer_features = customer_features.drop(list(vif_before[vif_before['VIF'] > 1000].Feature), axis=1)
customer_features.shape

(4466, 183)

In [ ]:
# re-examine VIF values after dropping extreme hight VIF features
vif_trimmed = pd.DataFrame(
    [variance_inflation_factor(customer_features.values, i) for i in range(customer_features.shape[1])],
    index=customer_features.columns,
    dtype=float,
).reset_index()
vif_trimmed.columns = ('Feature', 'VIF')

In [ ]:
print('After trimming extremely high VIF features:')
vif_trimmed

After trimming extremely high VIF features:


Feature     VIF
0                avg_event_params_engagement_time_msec  16.883
1                                avg_item_price_in_usd  86.707
2                                  avg_item_promotions  10.286
3                                    avg_item_quantity  45.344
4                               avg_item_refund_in_usd  25.289
5                             avg_item_refund_quantity  24.457
6                              avg_item_revenue_in_usd  98.243
7                    sum_ecommerce_total_item_quantity 448.286
8                       sum_total_return_item_quantity 199.463
9                sum_ecommerce_purchase_revenue_in_usd 270.271
10                   sum_ecommerce_refund_value_in_usd 191.970
11                      sum_ecommerce_tax_value_in_usd  46.821
12                          sum_ecommerce_unique_items 120.798
13               max_days_first_session_to_transaction   2.402
14                                max_user_ltv_revenue  25.507
15               max_event_params_engagement_time_msec   4.471
16                               max_item_price_in_usd  93.732
17                                 max_item_promotions  11.037
18                                   max_item_quantity  30.951
19                              max_item_refund_in_usd  48.847
20                            max_item_refund_quantity  39.501
21                             max_item_revenue_in_usd 125.057
22               sum_event_params_engagement_time_msec  40.446
23                               sum_item_price_in_usd 481.165
24                                 sum_item_promotions   9.890
25                                   sum_item_quantity 133.144
26                              sum_item_refund_in_usd 172.157
27                            sum_item_refund_quantity 196.698
28                             sum_item_revenue_in_usd 734.508
29                               sum_event_name_scroll 127.019
30                          sum_device_category_mobile  11.167
31                 sum_device_mobile_brand_name_Google 404.212
32                 sum_device_mobile_brand_name_Huawei   2.027
33                sum_device_mobile_brand_name_Samsung   3.188
34                 sum_device_mobile_brand_name_Xiaomi   3.344
35                 sum_device_mobile_model_name_Chrome 406.009
36             sum_device_mobile_model_name_ChromeBook  14.779
37          sum_device_web_info_browser_AndroidWebview   8.439
38                  sum_device_web_info_browser_Chrome 115.413
39                    sum_device_web_info_browser_Edge   2.138
40                  sum_device_web_info_browser_Safari  18.272
41                              sum_geo_country_Canada   4.600
42                              sum_geo_country_France   2.199
43                               sum_geo_country_India  15.245
44                               sum_geo_country_Other  98.908
45                               sum_geo_country_Spain  25.486
46                       sum_geo_country_UnitedKingdom   7.936
47                        sum_geo_country_UnitedStates  30.002
48                     sum_traffic_source_medium_Other   2.362
49                       sum_traffic_source_medium_cpc   1.271
50                   sum_traffic_source_medium_organic   3.954
51                  sum_traffic_source_medium_referral   7.038
52       sum_event_params_parent_page_CampusCollection   1.650
53   sum_event_params_parent_page_CheckoutConfirmation 176.010
54            sum_event_params_parent_page_EcoFriendly  27.025
55                   sum_event_params_parent_page_Home  17.936
56                    sum_event_params_parent_page_New  10.146
57                  sum_event_params_parent_page_Other  20.402
58          sum_event_params_parent_page_PaymentMethod 236.872
59                   sum_event_params_parent_page_Sale  87.625
60            sum_event_params_parent_page_ShopbyBrand 151.618
61           sum_event_params_parent_page_ShoppingCart  89.300
62             sum_event_params_parent_page_Stationery 179.034
63             

In [ ]:
vif_trimmed[vif_trimmed['VIF'] > 500].shape[0]

1

After removing extremely high VIF features, only 1 feature still has a VIF value over 500. <br>
Let's start by automatically removing features to get all features to a max VIF value of 10.

In [ ]:
# function to calculate VIF scores and loop through to drop col with max VIF each time until all are less than 10
def vif_and_drop(X, thresh=5.0):
    cols = X.columns
    features = np.arange(X.shape[1])
    dropped=True
    while dropped:
        dropped=False
        c = X[cols[features]].values
        vif = [variance_inflation_factor(c, ix) for ix in np.arange(c.shape[1])]

        maxloc = vif.index(max(vif))
        if max(vif) > thresh:
            print('dropping \'' + X[cols[features]].columns[maxloc] + '\' at index: ' + str(maxloc))
            features = np.delete(features, maxloc)
            dropped=True

    X = X[cols[features]]
    return X

In [ ]:
# call the function and update customer_features
customer_features = vif_and_drop(customer_features, thresh=10)

dropping 'sum_item_revenue_in_usd' at index: 28
dropping 'sum_ecommerce_total_item_quantity' at index: 7
dropping 'sum_device_mobile_model_name_Chrome' at index: 33
dropping 'pre_sum_event_name_page_view' at index: 92
dropping 'sum_item_child_category_Other' at index: 79
dropping 'pre_nunique_event_params_ga_sessions' at index: 87
dropping 'sum_event_params_parent_page_PaymentMethod' at index: 55
dropping 'sum_item_parent_category_ShopbyBrand' at index: 73
dropping 'sum_item_refund_quantity' at index: 26
dropping 'sum_ecommerce_purchase_revenue_in_usd' at index: 8
dropping 'pre_sum_device_mobile_brand_name_Apple' at index: 91
dropping 'sum_item_refund_in_usd' at index: 24
dropping 'pre_sum_event_params_parent_page_ShopbyBrand' at index: 128
dropping 'sum_item_parent_category_Stationery' at index: 70
dropping 'sum_event_params_child_page_Kids' at index: 58
dropping 'sum_event_params_parent_page_CheckoutConfirmation' at index: 47
dropping 'pre_sum_device_category_desktop' at index: 85
dr

In [ ]:
# check all final VIF values
vif_reduced = pd.DataFrame(
    [variance_inflation_factor(customer_features.values, i) for i in range(customer_features.shape[1])],
    index=customer_features.columns,
    dtype=float,
).reset_index()
vif_reduced.columns = ('Feature', 'VIF')

In [ ]:
print('All features VIF < 10:')
vif_reduced

All features VIF < 10:


Feature   VIF
0                avg_event_params_engagement_time_msec 7.136
1                                avg_item_price_in_usd 5.159
2                                  avg_item_promotions 7.040
3                               avg_item_refund_in_usd 3.208
4                       sum_total_return_item_quantity 5.154
5                    sum_ecommerce_refund_value_in_usd 6.960
6                max_days_first_session_to_transaction 1.957
7                max_event_params_engagement_time_msec 3.938
8                                  max_item_promotions 7.376
9                                    max_item_quantity 6.165
10                                 sum_item_promotions 4.553
11                                   sum_item_quantity 3.886
12                          sum_device_category_mobile 7.946
13                 sum_device_mobile_brand_name_Google 5.616
14                 sum_device_mobile_brand_name_Huawei 1.592
15                sum_device_mobile_brand_name_Samsung 2.656
16                 sum_device_mobile_brand_name_Xiaomi 1.922
17             sum_device_mobile_model_name_ChromeBook 1.556
18          sum_device_web_info_browser_AndroidWebview 3.887
19                    sum_device_web_info_browser_Edge 1.465
20                  sum_device_web_info_browser_Safari 5.268
21                              sum_geo_country_Canada 1.568
22                              sum_geo_country_France 1.471
23                               sum_geo_country_India 2.583
24                               sum_geo_country_Spain 8.021
25                       sum_geo_country_UnitedKingdom 2.552
26                        sum_geo_country_UnitedStates 2.724
27                     sum_traffic_source_medium_Other 1.714
28                       sum_traffic_source_medium_cpc 1.183
29                   sum_traffic_source_medium_organic 2.741
30                  sum_traffic_source_medium_referral 4.231
31       sum_event_params_parent_page_CampusCollection 1.337
32            sum_event_params_parent_page_EcoFriendly 1.502
33                   sum_event_params_parent_page_Home 1.953
34                    sum_event_params_parent_page_New 3.574
35                  sum_event_params_parent_page_Other 5.405
36                   sum_event_params_parent_page_Sale 1.753
37             sum_event_params_parent_page_Stationery 8.355
38                  sum_event_params_child_page_Google 6.509
39                    sum_event_params_child_page_Hats 3.132
40               sum_event_params_child_page_Notebooks 3.471
41                   sum_event_params_child_page_Other 2.329
42                   sum_event_params_child_page_Socks 2.191
43                sum_event_params_child_page_Stickers 1.800
44                 sum_event_params_child_page_Writing 9.281
45                 sum_event_params_child_page_YouTube 1.401
46                        sum_item_parent_category_New 4.471
47                      sum_item_parent_category_Other 1.401
48                      sum_item_child_category_Google 7.301
49                        sum_item_child_category_Kids 1.435
50                  sum_item_child_category_MensUnisex 4.624
51                      sum_item_child_category_Womens 2.363
52                sum_item_child_subcategory_Backpacks 1.403
53   sum_item_child_subcategory_ElectronicsAccessories 1.550
54                   sum_item_child_subcategory_Infant   NaN
55              sum_item_child_subcategory_MensTShirts   NaN
56             sum_item_child_subcategory_MugsTumblers   NaN
57                    sum_item_child_subcategory_Other   NaN
58             sum_item_child_subcategory_WaterBottles   NaN
59           pre_sum_event_params_engagement_time_msec 6.902
60                      pre_sum_event_name_add_to_cart 3.807
61                   pre_sum_event_name_begin_checkout 2.322
62                      pre_sum_event_name_select_item 3.155
63                      pre_sum_device_category_mobile 7.614
64             pre_sum_device_mobile_brand_name_Google 4.373
65             pre_su

In [ ]:
customer_features.shape

(4466, 125)

In [ ]:
df_customer_reduced = pd.concat([df_customer.iloc[:, :2], customer_features, df_customer.iloc[:, -1]], axis=1)
df_customer_reduced.head()

user_pseudo_id ecommerce_transaction_id  \
0   1219636.7573278529                   775080   
1  12291446.3815334458                   205482   
2  12291446.3815334458                   597612   
3  13756133.7762491194                   462859   
4   1415699.2356865981                   484939   

   avg_event_params_engagement_time_msec  avg_item_price_in_usd  \
0                               3754.163                  8.625   
1                               9148.439                 28.857   
2                              10078.185                 18.529   
3                               8643.410                 19.000   
4                               4372.630                 29.000   

   avg_item_promotions  avg_item_refund_in_usd  \
0                0.292                  10.125   
1                0.000                  28.857   
2                0.000                  18.529   
3                1.000                  38.000   
4                1.000                   0.000   

   sum_total_return_item_quantity  sum_ecommerce_refund_value_in_usd  \
0                           9.000                             70.000   
1                           7.000                            203.000   
2                          12.000                            244.000   
3                           4.000                             76.000   
4                           0.000                              0.000   

   max_days_first_session_to_transaction  \
0                                  0.000   
1                                 10.000   
2                                 10.000   
3                                  4.000   
4                                  0.000   

   max_event_params_engagement_time_msec  max_item_promotions  \
0                              95696.000                1.000   
1                             186421.000                0.000   
2                             307177.000                0.000   
3                             332905.000                1.000   
4                              58902.000                1.000   

   max_item_quantity  sum_item_promotions  sum_item_quantity  \
0              2.000                2.000              9.000   
1              1.000                0.000              7.000   
2              1.000                0.000             12.000   
3             10.000                2.000             19.000   
4              3.000                9.000             13.000   

   sum_device_category_mobile  sum_device_mobile_brand_name_Google  \
0                       2.000                                0.000   
1                       1.000                                0.000   
2                       2.000                                0.000   
3                       2.000                                0.000   
4                       1.000                                0.000   

   sum_device_mobile_brand_name_Huawei  sum_device_mobile_brand_name_Samsung  \
0                                0.000                                 0.000   
1                                0.000                                 0.000   
2                                0.000                                 0.000   
3                                0.000                                 0.000   
4                                0.000                                 0.000   

   sum_device_mobile_brand_name_Xiaomi  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                2.000   
4                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_web_info_browser_AndroidWebview  \
0                                       0.

We've reduced the number of columns from 223 to 125. All scores are now less than 10, so we can move on to the the transactions features.

In [ ]:
# create table schema to prevent errors when uploading data
schema = []

for col, dtype in zip(df_customer_reduced.dtypes.index, df_customer_reduced.dtypes.values):
    col_dict = {}
    col_dict['name'] = col

    if dtype == object:
        typ = 'STRING'
    elif dtype == bool:
        typ = 'BOOLEAN'
    elif dtype == 'datetime64[ns]':
        typ = 'DATETIME'
    elif dtype in (int, 'Int64', 'uint8'):
        typ = 'INTEGER'
    elif dtype == float:
        typ = 'FLOAT'
    else:
        print(dtype)
    col_dict['type'] = typ

    schema.append(col_dict)

In [ ]:
# df_customer_reduced.to_gbq(f'{project_name}.return_prediction_ga4.step_5_customer_reduced',
#                             project_id=project_name,
#                             if_exists='replace',
#                             location=region,
#                             chunksize=100_000,
#                             table_schema=schema)

# Load transaction data

In [ ]:
sql = f"""
SELECT *
FROM `{transaction_table_name}`;
"""
transaction_data = pandas_gbq.read_gbq(sql, project_id=project_name, location=region, use_bqstorage_api=True)

Downloading: 100%|██████████|


In [ ]:
# create a copy of the imported data to avoid re-importing if we need to revert to original table
df_transaction = transaction_data.copy()

In [ ]:
df_transaction.head()

user_pseudo_id transaction_date  transaction_ga_session_number  \
0  10111055.8768683862       2020-12-10                              1   
1   1019527.5799124267       2020-12-05                              1   
2   1021887.1151788384       2020-12-26                              1   
3   1026932.0858862293       2020-12-01                              1   
4   1026932.0858862293       2020-12-08                              3   

  ecommerce_transaction_id  sum_event_params_engagement_time_msec  \
0                   741471                                3324661   
1                     2105                                2966128   
2                    76937                                 918786   
3                   339943                                2387211   
4                    29460                                 674555   

   sum_event_params_session_engaged  user_ltv_revenue  \
0                               225            94.000   
1                               412            44.000   
2                               177            21.000   
3                               173           110.000   
4                               127           214.000   

   ecommerce_total_item_quantity  total_return_item_quantity  \
0                              3                           0   
1                              7                           0   
2                              2                           0   
3                              1                           0   
4                              1                           0   

   ecommerce_purchase_revenue_in_usd  ecommerce_refund_value_in_usd  \
0                             94.000                          0.000   
1                             44.000                          0.000   
2                             21.000                          0.000   
3                             55.000                          0.000   
4                             55.000                          0.000   

   ecommerce_tax_value_in_usd  ecommerce_unique_items  sum_item_price_in_usd  \
0                      10.000                       3                 94.000   
1                       5.000                       4                 32.000   
2                       2.000                       1                 10.000   
3                       4.000                       1                110.000   
4                       4.000                       1                 55.000   

   sum_item_promotions  sum_item_quantity  sum_item_refund_in_usd  \
0                    0                  3                   0.000   
1                    0                  7                   0.000   
2                    1                  2                   0.000   
3                    2                  2                   0.000   
4                    0                  1                   0.000   

   sum_item_refund_quantity  sum_item_revenue_in_usd  \
0                         0                   94.000   
1                         0                   44.000   
2                         0                   21.000   
3                         0                  110.000   
4                         0                   55.000   

   sum_event_name_add_to_cart  sum_event_name_begin_checkout  \
0                          48                              6   
1                          78                              8   
2                          48                              9   
3                          12                              2   
4                           0                              4   

   sum_event_name_page_view  sum_event_name_scroll  \
0                        26                     16   
1                        34                     21   
2                        15                     13   
3                        28                     19   
4                        25                     11   

   sum_event_name_select_item  sum_event_name_user

In [ ]:
df_transaction.shape

(4466, 225)

In [ ]:
# check for missing values
df_transaction.isna().sum().sum()

0

# Transaction VIF analysis

In [ ]:
# remove features related to refund/return since this is our target
transaction_features = df_transaction.drop([col for col in df_transaction.columns if ('refund' in col) or ('return' in col)], axis=1)
transaction_features.shape

(4466, 217)

In [ ]:
# remove id columns from analysis
transaction_features = transaction_features.drop(['user_pseudo_id', 'transaction_date', 'ecommerce_transaction_id', 'split'], axis=1)
transaction_features.head()

transaction_ga_session_number  sum_event_params_engagement_time_msec  \
0                              1                                3324661   
1                              1                                2966128   
2                              1                                 918786   
3                              1                                2387211   
4                              3                                 674555   

   sum_event_params_session_engaged  user_ltv_revenue  \
0                               225            94.000   
1                               412            44.000   
2                               177            21.000   
3                               173           110.000   
4                               127           214.000   

   ecommerce_total_item_quantity  ecommerce_purchase_revenue_in_usd  \
0                              3                             94.000   
1                              7                             44.000   
2                              2                             21.000   
3                              1                             55.000   
4                              1                             55.000   

   ecommerce_tax_value_in_usd  ecommerce_unique_items  sum_item_price_in_usd  \
0                      10.000                       3                 94.000   
1                       5.000                       4                 32.000   
2                       2.000                       1                 10.000   
3                       4.000                       1                110.000   
4                       4.000                       1                 55.000   

   sum_item_promotions  sum_item_quantity  sum_item_revenue_in_usd  \
0                    0                  3                   94.000   
1                    0                  7                   44.000   
2                    1                  2                   21.000   
3                    2                  2                  110.000   
4                    0                  1                   55.000   

   sum_event_name_add_to_cart  sum_event_name_begin_checkout  \
0                          48                              6   
1                          78                              8   
2                          48                              9   
3                          12                              2   
4                           0                              4   

   sum_event_name_page_view  sum_event_name_scroll  \
0                        26                     16   
1                        34                     21   
2                        15                     13   
3                        28                     19   
4                        25                     11   

   sum_event_name_select_item  sum_event_name_user_engagement  \
0                          12                              25   
1                           0                              33   
2                           0                              15   
3                           0                              28   
4                          27                              24   

   sum_event_name_view_item  sum_device_category_desktop  \
0                        84                            0   
1                       231                            0   
2                        72                            1   
3                        77                            1   
4                        27                            1   

   sum_device_category_mobile  sum_device_mobile_brand_name_Apple  \
0                           1                                   0   
1                           1                                   0   
2                           0                                   1   
3                           0                                   0   
4                           0                                   0   

   

In [ ]:
transaction_features.shape

(4466, 213)

In [ ]:
transaction_features = transaction_features.astype('float64')

In [ ]:
# first, let's take a look at VIF scores for all original features in the customer dataset
vif_before = pd.DataFrame(
    [variance_inflation_factor(transaction_features.values, i) for i in range(transaction_features.shape[1])],
    index=transaction_features.columns,
    dtype=float,
).reset_index()
vif_before.columns = ('Feature', 'VIF')

In [ ]:
print('Before feature selection:')
vif_before

Before feature selection:


Feature       VIF
0                        transaction_ga_session_number     6.085
1                sum_event_params_engagement_time_msec    18.369
2                     sum_event_params_session_engaged 10889.288
3                                     user_ltv_revenue     5.226
4                        ecommerce_total_item_quantity    78.908
5                    ecommerce_purchase_revenue_in_usd    60.065
6                           ecommerce_tax_value_in_usd    10.440
7                               ecommerce_unique_items    17.525
8                                sum_item_price_in_usd    44.492
9                                  sum_item_promotions     2.978
10                                   sum_item_quantity    79.922
11                             sum_item_revenue_in_usd   108.344
12                          sum_event_name_add_to_cart  1355.064
13                       sum_event_name_begin_checkout   158.864
14                            sum_event_name_page_view   607.975
15                               sum_event_name_scroll    44.174
16                          sum_event_name_select_item   178.599
17                      sum_event_name_user_engagement   405.124
18                            sum_event_name_view_item  4533.565
19                         sum_device_category_desktop   149.529
20                          sum_device_category_mobile    62.437
21                  sum_device_mobile_brand_name_Apple   565.468
22                 sum_device_mobile_brand_name_Google    33.605
23                 sum_device_mobile_brand_name_Huawei     1.777
24              sum_device_mobile_brand_name_Microsoft    25.925
25                sum_device_mobile_brand_name_Mozilla       inf
26                sum_device_mobile_brand_name_Samsung     3.108
27                 sum_device_mobile_brand_name_Xiaomi     1.851
28                 sum_device_mobile_model_name_Chrome    58.804
29             sum_device_mobile_model_name_ChromeBook    12.981
30                   sum_device_mobile_model_name_Edge    93.127
31                sum_device_mobile_model_name_Firefox       inf
32                 sum_device_mobile_model_name_Safari   297.733
33                   sum_device_mobile_model_name_iPad    17.607
34                 sum_device_mobile_model_name_iPhone   276.147
35          sum_device_web_info_browser_AndroidWebview     2.638
36                  sum_device_web_info_browser_Chrome    47.822
37                    sum_device_web_info_browser_Edge    70.634
38                 sum_device_web_info_browser_Firefox     5.018
39                  sum_device_web_info_browser_Safari    23.044
40                              sum_geo_country_Canada    14.802
41                              sum_geo_country_France     5.157
42                               sum_geo_country_India    16.245
43                               sum_geo_country_Other    51.509
44                               sum_geo_country_Spain     5.575
45                       sum_geo_country_UnitedKingdom     6.550
46                        sum_geo_country_UnitedStates    71.831
47                     sum_traffic_source_medium_Other     1.405
48                       sum_traffic_source_medium_cpc     1.138
49                   sum_traffic_source_medium_organic     1.908
50                  sum_traffic_source_medium_referral     1.708
51                sum_event_params_parent_page_Apparel   455.338
52       sum_event_params_parent_page_CampusCollection    26.177
53   sum_event_params_parent_page_CheckoutConfirmation    17.650
54   sum_event_params_parent_page_CheckoutYourInfor...   105.306
55            sum_event_params_parent_page_EcoFriendly    13.337
56                   sum_event_params_parent_page_Home   122.037
57              sum_event_params_parent_page_Lifestyle   200.319
58                    sum_event_params_parent_page_New    17.790
59                  sum_event_params_parent_page_Other     2.983
60          sum_event_params_parent_page_PaymentMethod    11.549
61               

In [ ]:
vif_before[vif_before['VIF'] > 5].shape[0]

167

Again, since so many of our features have such high VIF values, let's start by dropping the features with a VIF value over 500 and recalculate VIF.

In [ ]:
# drop 22 features with VIF over 500
transaction_features = transaction_features.drop(list(vif_before[vif_before['VIF'] > 500].Feature), axis=1)
transaction_features.shape

(4466, 191)

In [ ]:
# re-examine VIF values after dropping extreme hight VIF features
vif_trimmed = pd.DataFrame(
    [variance_inflation_factor(transaction_features.values, i) for i in range(transaction_features.shape[1])],
    index=transaction_features.columns,
    dtype=float,
).reset_index()
vif_trimmed.columns = ('Feature', 'VIF')

In [ ]:
print('After trimming extremely high VIF features:')
vif_trimmed

After trimming extremely high VIF features:


Feature     VIF
0                        transaction_ga_session_number   5.994
1                sum_event_params_engagement_time_msec  17.155
2                                     user_ltv_revenue   5.175
3                        ecommerce_total_item_quantity  78.232
4                    ecommerce_purchase_revenue_in_usd  59.434
5                           ecommerce_tax_value_in_usd  10.403
6                               ecommerce_unique_items  16.900
7                                sum_item_price_in_usd  43.652
8                                  sum_item_promotions   2.942
9                                    sum_item_quantity  79.052
10                             sum_item_revenue_in_usd 107.337
11                       sum_event_name_begin_checkout  68.640
12                               sum_event_name_scroll  18.617
13                          sum_event_name_select_item   2.911
14                      sum_event_name_user_engagement  31.959
15                         sum_device_category_desktop 134.295
16                          sum_device_category_mobile  61.943
17                 sum_device_mobile_brand_name_Google  33.320
18                 sum_device_mobile_brand_name_Huawei   1.760
19              sum_device_mobile_brand_name_Microsoft  25.521
20                sum_device_mobile_brand_name_Samsung   3.037
21                 sum_device_mobile_brand_name_Xiaomi   1.838
22                 sum_device_mobile_model_name_Chrome  50.958
23             sum_device_mobile_model_name_ChromeBook  11.393
24                   sum_device_mobile_model_name_Edge  92.810
25                 sum_device_mobile_model_name_Safari  19.495
26                   sum_device_mobile_model_name_iPad   3.152
27                 sum_device_mobile_model_name_iPhone  11.146
28          sum_device_web_info_browser_AndroidWebview   2.593
29                  sum_device_web_info_browser_Chrome  46.699
30                    sum_device_web_info_browser_Edge  70.501
31                 sum_device_web_info_browser_Firefox   2.717
32                  sum_device_web_info_browser_Safari  22.708
33                              sum_geo_country_Canada  14.653
34                              sum_geo_country_France   5.119
35                               sum_geo_country_India  16.114
36                               sum_geo_country_Other  50.931
37                               sum_geo_country_Spain   5.517
38                       sum_geo_country_UnitedKingdom   6.508
39                        sum_geo_country_UnitedStates  71.113
40                     sum_traffic_source_medium_Other   1.395
41                       sum_traffic_source_medium_cpc   1.133
42                   sum_traffic_source_medium_organic   1.901
43                  sum_traffic_source_medium_referral   1.695
44                sum_event_params_parent_page_Apparel  24.476
45       sum_event_params_parent_page_CampusCollection   1.526
46   sum_event_params_parent_page_CheckoutConfirmation  10.999
47   sum_event_params_parent_page_CheckoutYourInfor...  99.143
48            sum_event_params_parent_page_EcoFriendly   6.689
49                   sum_event_params_parent_page_Home   4.834
50              sum_event_params_parent_page_Lifestyle  16.177
51                    sum_event_params_parent_page_New  13.804
52                  sum_event_params_parent_page_Other   2.157
53          sum_event_params_parent_page_PaymentMethod  10.278
54                   sum_event_params_parent_page_Sale   2.105
55            sum_event_params_parent_page_ShopbyBrand  86.074
56           sum_event_params_parent_page_ShoppingCart   8.439
57             sum_event_params_parent_page_Stationery  52.567
58                    sum_event_params_child_page_Bags  26.582
59               sum_event_params_child_page_Drinkware  21.581
60                  sum_event_params_child_page_Google  37.470
61                    sum_event_params_child_page_Hats  10.630
62                    sum_event_params_child_page_Kids  11.554
63             

In [ ]:
vif_trimmed[vif_trimmed['VIF'] > 200].shape[0]

3

After removing extremely high VIF features, only 3 features still has a VIF value over 200. <br>
Let's start by automatically removing features to get all features to a max VIF value of 10.

In [ ]:
# call the function and update customer_features
transaction_features = vif_and_drop(transaction_features, thresh=10)

dropping 'pre_sum_event_name_page_view' at index: 109
dropping 'pre_nunique_event_params_ga_sessions' at index: 101
dropping 'pre_sum_device_mobile_brand_name_Apple' at index: 113
dropping 'pre_sum_device_category_desktop' at index: 111
dropping 'sum_item_child_category_Other' at index: 81
dropping 'pre_sum_event_params_parent_page_ShopbyBrand' at index: 149
dropping 'sum_item_revenue_in_usd' at index: 10
dropping 'sum_event_params_parent_page_CheckoutYourInformation' at index: 46
dropping 'sum_device_category_desktop' at index: 14
dropping 'sum_device_mobile_model_name_Edge' at index: 22
dropping 'pre_sum_item_child_category_Other' at index: 170
dropping 'sum_event_params_parent_page_ShopbyBrand' at index: 51
dropping 'ecommerce_total_item_quantity' at index: 3
dropping 'pre_sum_device_mobile_model_name_Edge' at index: 112
dropping 'pre_sum_device_mobile_model_name_Chrome' at index: 110
dropping 'pre_sum_event_params_child_page_SmallGoods' at index: 151
dropping 'pre_sum_event_params_

In [ ]:
# check all final VIF values
vif_reduced = pd.DataFrame(
    [variance_inflation_factor(transaction_features.values, i) for i in range(transaction_features.shape[1])],
    index=transaction_features.columns,
    dtype=float,
).reset_index()
vif_reduced.columns = ('Feature', 'VIF')

In [ ]:
print('All features VIF < 10:')
vif_reduced

All features VIF < 10:


Feature   VIF
0                        transaction_ga_session_number 5.783
1                                     user_ltv_revenue 4.541
2                           ecommerce_tax_value_in_usd 6.158
3                                sum_item_price_in_usd 8.343
4                                  sum_item_promotions 2.613
5                                    sum_item_quantity 3.816
6                        sum_event_name_begin_checkout 4.143
7                           sum_event_name_select_item 2.224
8                           sum_device_category_mobile 7.963
9                  sum_device_mobile_brand_name_Google 4.632
10                 sum_device_mobile_brand_name_Huawei 1.661
11              sum_device_mobile_brand_name_Microsoft 1.603
12                sum_device_mobile_brand_name_Samsung 2.457
13                 sum_device_mobile_brand_name_Xiaomi 1.715
14             sum_device_mobile_model_name_ChromeBook 1.771
15                 sum_device_mobile_model_name_Safari 3.505
16                   sum_device_mobile_model_name_iPad 1.491
17                 sum_device_mobile_model_name_iPhone 7.210
18          sum_device_web_info_browser_AndroidWebview 1.713
19                 sum_device_web_info_browser_Firefox 1.402
20                  sum_device_web_info_browser_Safari 5.667
21                              sum_geo_country_Canada 1.833
22                              sum_geo_country_France 1.651
23                               sum_geo_country_India 1.920
24                               sum_geo_country_Other 2.676
25                               sum_geo_country_Spain 1.790
26                       sum_geo_country_UnitedKingdom 1.793
27                     sum_traffic_source_medium_Other 1.357
28                       sum_traffic_source_medium_cpc 1.110
29                   sum_traffic_source_medium_organic 1.808
30                  sum_traffic_source_medium_referral 1.631
31       sum_event_params_parent_page_CampusCollection 1.362
32   sum_event_params_parent_page_CheckoutConfirmation 8.706
33            sum_event_params_parent_page_EcoFriendly 1.231
34                   sum_event_params_parent_page_Home 1.932
35              sum_event_params_parent_page_Lifestyle 9.672
36                    sum_event_params_parent_page_New 1.268
37                  sum_event_params_parent_page_Other 1.886
38          sum_event_params_parent_page_PaymentMethod 6.686
39                   sum_event_params_parent_page_Sale 1.492
40           sum_event_params_parent_page_ShoppingCart 5.295
41                    sum_event_params_child_page_Bags 2.638
42                  sum_event_params_child_page_Google 5.154
43                    sum_event_params_child_page_Hats 1.344
44                    sum_event_params_child_page_Kids 9.505
45               sum_event_params_child_page_Notebooks 4.345
46                   sum_event_params_child_page_Other 2.702
47                   sum_event_params_child_page_Socks 1.348
48                sum_event_params_child_page_Stickers 1.340
49                  sum_event_params_child_page_Womens 1.415
50                 sum_event_params_child_page_Writing 2.010
51                 sum_event_params_child_page_YouTube 1.181
52                      sum_item_parent_category_Other 1.473
53                 sum_item_parent_category_Stationery 5.185
54                   sum_item_child_category_Drinkware 4.630
55                      sum_item_child_category_Google 5.379
56                        sum_item_child_category_Kids 9.930
57                  sum_item_child_category_MensUnisex 1.854
58                  sum_item_child_category_SmallGoods 2.461
59                sum_item_child_subcategory_Backpacks 1.032
60   sum_item_child_subcategory_ElectronicsAccessories 1.321
61                   sum_item_child_subcategory_Infant 1.015
62              sum_item_child_subcategory_MensTShirts 1.135
63             sum_item_child_subcategory_MugsTumblers 1.196
64                    sum_item_child_subcategory_Other 1.039
65             sum_it

In [ ]:
transaction_features.shape

(4466, 137)

In [ ]:
df_transaction_reduced = pd.concat([df_transaction.loc[:, ['user_pseudo_id', 'ecommerce_transaction_id']],
                                    transaction_features,
                                    df_transaction.iloc[:, -1]],
                                   axis=1)
df_transaction_reduced.head()

user_pseudo_id ecommerce_transaction_id  \
0  10111055.8768683862                   741471   
1   1019527.5799124267                     2105   
2   1021887.1151788384                    76937   
3   1026932.0858862293                   339943   
4   1026932.0858862293                    29460   

   transaction_ga_session_number  user_ltv_revenue  \
0                          1.000            94.000   
1                          1.000            44.000   
2                          1.000            21.000   
3                          1.000           110.000   
4                          3.000           214.000   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      10.000                 94.000                0.000   
1                       5.000                 32.000                0.000   
2                       2.000                 10.000                1.000   
3                       4.000                110.000                2.000   
4                       4.000                 55.000                0.000   

   sum_item_quantity  sum_event_name_begin_checkout  \
0              3.000                          6.000   
1              7.000                          8.000   
2              2.000                          9.000   
3              2.000                          2.000   
4              1.000                          4.000   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      12.000                       1.000   
1                       0.000                       1.000   
2                       0.000                       0.000   
3                       0.000                       0.000   
4                      27.000                       0.000   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                                0.000                                0.000   
1                                0.000                                0.000   
2                                0.000                                0.000   
3                                1.000                                0.000   
4                                1.000                                0.000   

   sum_device_mobile_brand_name_Microsoft  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3                                   0.000   
4                                   0.000   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                 0.000                                0.000   
1                                 1.000                                0.000   
2                                 0.000                                0.000   
3                                 0.000                                0.000   
4                                 0.000                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                                0.000                              0.000   
1                                0.000                              0.000   
2                                1.000                              0.000   
3                                0.000                              0.000   
4                                0.000                              0.000   

   sum_device_mobile_model_name_iPhone  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                0.000   
4                                0.000

We've reduced the number of columns from 213 to 137. All scores are now less than 10, so we can move on to the the transactions features.

In [ ]:
# create table schema to prevent errors when uploading data
schema = []

for col, dtype in zip(df_transaction_reduced.dtypes.index, df_transaction_reduced.dtypes.values):
    col_dict = {}
    col_dict['name'] = col

    if dtype == object:
        typ = 'STRING'
    elif dtype == bool:
        typ = 'BOOLEAN'
    elif dtype == 'datetime64[ns]':
        typ = 'DATETIME'
    elif dtype in (int, 'Int64', 'uint8'):
        typ = 'INTEGER'
    elif dtype == float:
        typ = 'FLOAT'
    else:
        print(dtype)
    col_dict['type'] = typ

    schema.append(col_dict)

In [ ]:
# df_transaction_reduced.to_gbq(f'{project_name}.return_prediction_ga4.step_5_transaction_reduced',
#                                 project_id=project_name,
#                                 if_exists='replace',
#                                 location=region,
#                                 chunksize=100_000,
#                                 table_schema=schema)

# Join transaction and customer data

In [ ]:
df_customer_reduced.columns = [f'historical_{col}' for col in df_customer_reduced.columns]

In [ ]:
df_merge = pd.merge(df_transaction_reduced.iloc[:, :-1],
                    df_customer_reduced,
                    how='left',
                    left_on=['user_pseudo_id', 'ecommerce_transaction_id'],
                    right_on=['historical_user_pseudo_id', 'historical_ecommerce_transaction_id'])
df_merge.head()

user_pseudo_id ecommerce_transaction_id  \
0  10111055.8768683862                   741471   
1   1019527.5799124267                     2105   
2   1021887.1151788384                    76937   
3   1026932.0858862293                   339943   
4   1026932.0858862293                    29460   

   transaction_ga_session_number  user_ltv_revenue  \
0                          1.000            94.000   
1                          1.000            44.000   
2                          1.000            21.000   
3                          1.000           110.000   
4                          3.000           214.000   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      10.000                 94.000                0.000   
1                       5.000                 32.000                0.000   
2                       2.000                 10.000                1.000   
3                       4.000                110.000                2.000   
4                       4.000                 55.000                0.000   

   sum_item_quantity  sum_event_name_begin_checkout  \
0              3.000                          6.000   
1              7.000                          8.000   
2              2.000                          9.000   
3              2.000                          2.000   
4              1.000                          4.000   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      12.000                       1.000   
1                       0.000                       1.000   
2                       0.000                       0.000   
3                       0.000                       0.000   
4                      27.000                       0.000   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                                0.000                                0.000   
1                                0.000                                0.000   
2                                0.000                                0.000   
3                                1.000                                0.000   
4                                1.000                                0.000   

   sum_device_mobile_brand_name_Microsoft  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3                                   0.000   
4                                   0.000   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                 0.000                                0.000   
1                                 1.000                                0.000   
2                                 0.000                                0.000   
3                                 0.000                                0.000   
4                                 0.000                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                                0.000                              0.000   
1                                0.000                              0.000   
2                                1.000                              0.000   
3                                0.000                              0.000   
4                                0.000                              0.000   

   sum_device_mobile_model_name_iPhone  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                0.000   
4                                0.000

In [ ]:
# drop repeated columns
df_merge = df_merge.drop(['historical_user_pseudo_id', 'historical_ecommerce_transaction_id'], axis=1)
df_merge = df_merge.rename({'historical_split': 'split'}, axis=1)
df_merge.shape

(4466, 265)

# Joint VIF analysis

In [ ]:
# remove id columns from analysis
merged_features = df_merge.drop(['user_pseudo_id', 'ecommerce_transaction_id', 'split'], axis=1)
merged_features.head()

transaction_ga_session_number  user_ltv_revenue  \
0                          1.000            94.000   
1                          1.000            44.000   
2                          1.000            21.000   
3                          1.000           110.000   
4                          3.000           214.000   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      10.000                 94.000                0.000   
1                       5.000                 32.000                0.000   
2                       2.000                 10.000                1.000   
3                       4.000                110.000                2.000   
4                       4.000                 55.000                0.000   

   sum_item_quantity  sum_event_name_begin_checkout  \
0              3.000                          6.000   
1              7.000                          8.000   
2              2.000                          9.000   
3              2.000                          2.000   
4              1.000                          4.000   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      12.000                       1.000   
1                       0.000                       1.000   
2                       0.000                       0.000   
3                       0.000                       0.000   
4                      27.000                       0.000   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                                0.000                                0.000   
1                                0.000                                0.000   
2                                0.000                                0.000   
3                                1.000                                0.000   
4                                1.000                                0.000   

   sum_device_mobile_brand_name_Microsoft  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3                                   0.000   
4                                   0.000   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                 0.000                                0.000   
1                                 1.000                                0.000   
2                                 0.000                                0.000   
3                                 0.000                                0.000   
4                                 0.000                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                                0.000                              0.000   
1                                0.000                              0.000   
2                                1.000                              0.000   
3                                0.000                              0.000   
4                                0.000                              0.000   

   sum_device_mobile_model_name_iPhone  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                0.000   
4                                0.000   

   sum_device_web_info_browser_AndroidWebview  \
0                                       0.000   
1                                       0.000   
2                                       0.000   
3                                       0.000   
4                                       0.000   

  

In [ ]:
merged_features.shape

(4466, 262)

In [ ]:
# first, let's take a look at VIF scores for all reduced features across customer and transaction
vif_before = pd.DataFrame(
    [variance_inflation_factor(merged_features.values, i) for i in range(merged_features.shape[1])],
    index=merged_features.columns,
    dtype=float,
).reset_index()
vif_before.columns = ('Feature', 'VIF')

In [ ]:
print('Before feature selection:')
vif_before

Before feature selection:


Feature       VIF
0                        transaction_ga_session_number     7.423
1                                     user_ltv_revenue     6.918
2                           ecommerce_tax_value_in_usd     6.405
3                                sum_item_price_in_usd     8.993
4                                  sum_item_promotions     2.683
5                                    sum_item_quantity     3.948
6                        sum_event_name_begin_checkout     4.500
7                           sum_event_name_select_item     2.346
8                           sum_device_category_mobile     9.046
9                  sum_device_mobile_brand_name_Google     5.025
10                 sum_device_mobile_brand_name_Huawei     1.729
11              sum_device_mobile_brand_name_Microsoft     1.703
12                sum_device_mobile_brand_name_Samsung     2.684
13                 sum_device_mobile_brand_name_Xiaomi     1.894
14             sum_device_mobile_model_name_ChromeBook     1.826
15                 sum_device_mobile_model_name_Safari     3.768
16                   sum_device_mobile_model_name_iPad     1.585
17                 sum_device_mobile_model_name_iPhone    10.451
18          sum_device_web_info_browser_AndroidWebview     1.809
19                 sum_device_web_info_browser_Firefox     1.450
20                  sum_device_web_info_browser_Safari     6.900
21                              sum_geo_country_Canada     1.933
22                              sum_geo_country_France     1.729
23                               sum_geo_country_India     2.032
24                               sum_geo_country_Other     2.834
25                               sum_geo_country_Spain     2.060
26                       sum_geo_country_UnitedKingdom     1.924
27                     sum_traffic_source_medium_Other     1.398
28                       sum_traffic_source_medium_cpc     1.119
29                   sum_traffic_source_medium_organic     1.873
30                  sum_traffic_source_medium_referral     1.690
31       sum_event_params_parent_page_CampusCollection     1.413
32   sum_event_params_parent_page_CheckoutConfirmation     9.375
33            sum_event_params_parent_page_EcoFriendly     1.246
34                   sum_event_params_parent_page_Home     2.053
35              sum_event_params_parent_page_Lifestyle    10.056
36                    sum_event_params_parent_page_New     1.297
37                  sum_event_params_parent_page_Other     1.967
38          sum_event_params_parent_page_PaymentMethod     6.874
39                   sum_event_params_parent_page_Sale     1.521
40           sum_event_params_parent_page_ShoppingCart     5.394
41                    sum_event_params_child_page_Bags     2.719
42                  sum_event_params_child_page_Google     5.274
43                    sum_event_params_child_page_Hats     1.403
44                    sum_event_params_child_page_Kids     9.972
45               sum_event_params_child_page_Notebooks     4.513
46                   sum_event_params_child_page_Other     2.761
47                   sum_event_params_child_page_Socks     1.423
48                sum_event_params_child_page_Stickers     1.376
49                  sum_event_params_child_page_Womens     1.456
50                 sum_event_params_child_page_Writing     2.119
51                 sum_event_params_child_page_YouTube     1.221
52                      sum_item_parent_category_Other     1.498
53                 sum_item_parent_category_Stationery     5.664
54                   sum_item_child_category_Drinkware     4.839
55                      sum_item_child_category_Google     5.442
56                        sum_item_child_category_Kids    10.466
57                  sum_item_child_category_MensUnisex     1.935
58                  sum_item_child_category_SmallGoods     2.526
59                sum_item_child_subcategory_Backpacks     1.035
60   sum_item_child_subcategory_ElectronicsAccessories     1.347
61               

In [ ]:
vif_before[vif_before['VIF'] > 5].shape[0]

176

Again, since so many of our features have such high VIF values, let's start by dropping the features with a VIF value over 100 and recalculate VIF.

In [ ]:
# drop 48 features with VIF over 100
merged_features = merged_features.drop(list(vif_before[vif_before['VIF'] > 100].Feature), axis=1)
merged_features.shape

(4466, 214)

In [ ]:
# re-examine VIF values after dropping extreme hight VIF features
vif_trimmed = pd.DataFrame(
    [variance_inflation_factor(merged_features.values, i) for i in range(merged_features.shape[1])],
    index=merged_features.columns,
    dtype=float,
).reset_index()
vif_trimmed.columns = ('Feature', 'VIF')

In [ ]:
print('After trimming extremely high VIF features:')
vif_trimmed

After trimming extremely high VIF features:


Feature    VIF
0                        transaction_ga_session_number  7.360
1                                     user_ltv_revenue  6.537
2                           ecommerce_tax_value_in_usd  6.244
3                                sum_item_price_in_usd  8.940
4                                  sum_item_promotions  2.663
5                                    sum_item_quantity  3.911
6                        sum_event_name_begin_checkout  4.327
7                           sum_event_name_select_item  2.320
8                           sum_device_category_mobile  7.344
9                  sum_device_mobile_brand_name_Google  4.909
10                 sum_device_mobile_brand_name_Huawei  1.319
11              sum_device_mobile_brand_name_Microsoft  1.680
12                sum_device_mobile_brand_name_Samsung  2.438
13                 sum_device_mobile_brand_name_Xiaomi  1.810
14             sum_device_mobile_model_name_ChromeBook  1.808
15                 sum_device_mobile_model_name_Safari  3.642
16                   sum_device_mobile_model_name_iPad  1.525
17                 sum_device_mobile_model_name_iPhone  8.068
18          sum_device_web_info_browser_AndroidWebview  1.761
19                 sum_device_web_info_browser_Firefox  1.438
20                  sum_device_web_info_browser_Safari  4.768
21                              sum_geo_country_Canada  1.916
22                              sum_geo_country_France  1.170
23                               sum_geo_country_India  2.003
24                               sum_geo_country_Other  2.771
25                               sum_geo_country_Spain  2.011
26                       sum_geo_country_UnitedKingdom  1.913
27                     sum_traffic_source_medium_Other  1.385
28                       sum_traffic_source_medium_cpc  1.111
29                   sum_traffic_source_medium_organic  1.850
30                  sum_traffic_source_medium_referral  1.668
31       sum_event_params_parent_page_CampusCollection  1.406
32   sum_event_params_parent_page_CheckoutConfirmation  9.199
33            sum_event_params_parent_page_EcoFriendly  1.237
34                   sum_event_params_parent_page_Home  2.021
35              sum_event_params_parent_page_Lifestyle  9.928
36                    sum_event_params_parent_page_New  1.290
37                  sum_event_params_parent_page_Other  1.929
38          sum_event_params_parent_page_PaymentMethod  6.760
39                   sum_event_params_parent_page_Sale  1.506
40           sum_event_params_parent_page_ShoppingCart  5.339
41                    sum_event_params_child_page_Bags  2.698
42                  sum_event_params_child_page_Google  5.231
43                    sum_event_params_child_page_Hats  1.391
44                    sum_event_params_child_page_Kids  9.790
45               sum_event_params_child_page_Notebooks  4.328
46                   sum_event_params_child_page_Other  2.622
47                   sum_event_params_child_page_Socks  1.413
48                sum_event_params_child_page_Stickers  1.349
49                  sum_event_params_child_page_Womens  1.448
50                 sum_event_params_child_page_Writing  2.074
51                 sum_event_params_child_page_YouTube  1.202
52                      sum_item_parent_category_Other  1.462
53                 sum_item_parent_category_Stationery  5.404
54                   sum_item_child_category_Drinkware  4.795
55                      sum_item_child_category_Google  5.408
56                        sum_item_child_category_Kids 10.257
57                  sum_item_child_category_MensUnisex  1.911
58                  sum_item_child_category_SmallGoods  2.461
59                sum_item_child_subcategory_Backpacks  1.034
60   sum_item_child_subcategory_ElectronicsAccessories  1.339
61                   sum_item_child_subcategory_Infant  1.016
62              sum_item_child_subcategory_MensTShirts  1.102
63             sum_item_child_subcategory_MugsTumblers  1.201
64               

In [ ]:
vif_trimmed[vif_trimmed['VIF'] > 50].shape[0]

15

After removing extremely high VIF features, 15 features still has a VIF value over 50. <br>
Let's start by automatically removing features to get all features to a max VIF value of 10.

In [ ]:
# call the function and update merged_features
merged_features = vif_and_drop(merged_features, thresh=10)

dropping 'historical_pre_sum_event_params_child_page_Socks' at index: 201
dropping 'historical_pre_sum_device_mobile_model_name_ChromeBook' at index: 178
dropping 'historical_pre_sum_event_params_child_page_Hats' at index: 198
dropping 'historical_pre_sum_geo_country_Other' at index: 184
dropping 'historical_pre_sum_item_child_category_Bags' at index: 201
dropping 'historical_pre_sum_geo_country_India' at index: 183
dropping 'historical_pre_sum_event_params_parent_page_Sale' at index: 195
dropping 'historical_pre_max_event_params_ga_session_number' at index: 203
dropping 'historical_pre_sum_event_params_parent_page_EcoFriendly' at index: 191
dropping 'historical_pre_sum_event_params_parent_page_PaymentMethod' at index: 193
dropping 'historical_pre_sum_traffic_source_medium_referral' at index: 188
dropping 'historical_pre_sum_device_mobile_brand_name_Google' at index: 174
dropping 'historical_pre_sum_device_mobile_model_name_iPad' at index: 177
dropping 'historical_pre_sum_geo_country_C

In [ ]:
# check all final VIF values
vif_reduced = pd.DataFrame(
    [variance_inflation_factor(merged_features.values, i) for i in range(merged_features.shape[1])],
    index=merged_features.columns,
    dtype=float,
).reset_index()
vif_reduced.columns = ('Feature', 'VIF')

In [ ]:
print('All features VIF < 10:')
vif_reduced

All features VIF < 10:


Feature   VIF
0                        transaction_ga_session_number 6.359
1                                     user_ltv_revenue 6.384
2                           ecommerce_tax_value_in_usd 6.141
3                                sum_item_price_in_usd 5.886
4                                  sum_item_promotions 2.644
5                                    sum_item_quantity 3.881
6                        sum_event_name_begin_checkout 4.303
7                           sum_event_name_select_item 2.258
8                           sum_device_category_mobile 7.194
9                  sum_device_mobile_brand_name_Google 4.690
10                 sum_device_mobile_brand_name_Huawei 1.312
11              sum_device_mobile_brand_name_Microsoft 1.662
12                sum_device_mobile_brand_name_Samsung 2.406
13                 sum_device_mobile_brand_name_Xiaomi 1.781
14             sum_device_mobile_model_name_ChromeBook 1.797
15                 sum_device_mobile_model_name_Safari 3.465
16                   sum_device_mobile_model_name_iPad 1.472
17                 sum_device_mobile_model_name_iPhone 7.945
18          sum_device_web_info_browser_AndroidWebview 1.749
19                 sum_device_web_info_browser_Firefox 1.413
20                  sum_device_web_info_browser_Safari 4.750
21                              sum_geo_country_Canada 1.891
22                              sum_geo_country_France 1.169
23                               sum_geo_country_India 1.997
24                               sum_geo_country_Other 2.743
25                               sum_geo_country_Spain 1.988
26                       sum_geo_country_UnitedKingdom 1.896
27                     sum_traffic_source_medium_Other 1.372
28                       sum_traffic_source_medium_cpc 1.108
29                   sum_traffic_source_medium_organic 1.836
30                  sum_traffic_source_medium_referral 1.642
31       sum_event_params_parent_page_CampusCollection 1.396
32   sum_event_params_parent_page_CheckoutConfirmation 8.949
33            sum_event_params_parent_page_EcoFriendly 1.233
34                   sum_event_params_parent_page_Home 1.877
35              sum_event_params_parent_page_Lifestyle 9.856
36                    sum_event_params_parent_page_New 1.284
37                  sum_event_params_parent_page_Other 1.921
38          sum_event_params_parent_page_PaymentMethod 6.665
39                   sum_event_params_parent_page_Sale 1.496
40           sum_event_params_parent_page_ShoppingCart 5.250
41                    sum_event_params_child_page_Bags 2.641
42                  sum_event_params_child_page_Google 5.145
43                    sum_event_params_child_page_Hats 1.372
44                    sum_event_params_child_page_Kids 1.320
45               sum_event_params_child_page_Notebooks 4.285
46                   sum_event_params_child_page_Other 2.598
47                   sum_event_params_child_page_Socks 1.399
48                sum_event_params_child_page_Stickers 1.342
49                  sum_event_params_child_page_Womens 1.433
50                 sum_event_params_child_page_Writing 2.048
51                 sum_event_params_child_page_YouTube 1.189
52                      sum_item_parent_category_Other 1.446
53                 sum_item_parent_category_Stationery 5.330
54                   sum_item_child_category_Drinkware 4.749
55                      sum_item_child_category_Google 5.341
56                  sum_item_child_category_MensUnisex 1.840
57                  sum_item_child_category_SmallGoods 2.434
58                sum_item_child_subcategory_Backpacks 1.032
59   sum_item_child_subcategory_ElectronicsAccessories 1.335
60                   sum_item_child_subcategory_Infant 1.014
61              sum_item_child_subcategory_MensTShirts 1.100
62             sum_item_child_subcategory_MugsTumblers 1.199
63                    sum_item_child_subcategory_Other 1.038
64             sum_item_child_subcategory_WaterBottles 1.009
65               max_

In [ ]:
merged_features.shape

(4466, 181)

In [ ]:
df_merge_reduced = pd.concat([df_merge.loc[:, ['user_pseudo_id', 'ecommerce_transaction_id']],
                              merged_features,
                              df_merge.iloc[:, -1]],
                             axis=1)
df_merge_reduced.head()

user_pseudo_id ecommerce_transaction_id  \
0  10111055.8768683862                   741471   
1   1019527.5799124267                     2105   
2   1021887.1151788384                    76937   
3   1026932.0858862293                   339943   
4   1026932.0858862293                    29460   

   transaction_ga_session_number  user_ltv_revenue  \
0                          1.000            94.000   
1                          1.000            44.000   
2                          1.000            21.000   
3                          1.000           110.000   
4                          3.000           214.000   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      10.000                 94.000                0.000   
1                       5.000                 32.000                0.000   
2                       2.000                 10.000                1.000   
3                       4.000                110.000                2.000   
4                       4.000                 55.000                0.000   

   sum_item_quantity  sum_event_name_begin_checkout  \
0              3.000                          6.000   
1              7.000                          8.000   
2              2.000                          9.000   
3              2.000                          2.000   
4              1.000                          4.000   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      12.000                       1.000   
1                       0.000                       1.000   
2                       0.000                       0.000   
3                       0.000                       0.000   
4                      27.000                       0.000   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                                0.000                                0.000   
1                                0.000                                0.000   
2                                0.000                                0.000   
3                                1.000                                0.000   
4                                1.000                                0.000   

   sum_device_mobile_brand_name_Microsoft  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3                                   0.000   
4                                   0.000   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                 0.000                                0.000   
1                                 1.000                                0.000   
2                                 0.000                                0.000   
3                                 0.000                                0.000   
4                                 0.000                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                                0.000                              0.000   
1                                0.000                              0.000   
2                                1.000                              0.000   
3                                0.000                              0.000   
4                                0.000                              0.000   

   sum_device_mobile_model_name_iPhone  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                0.000   
4                                0.000

We've reduced the number of columns from 259 to 181. All scores are now less than 10, so we can move on to the next step. <br>
At this point, we should also re-attach our target column, `ecommerce_refund_value_in_usd`.

In [ ]:
df_merge_final = pd.merge(
    df_merge_reduced,
    df_transaction[['user_pseudo_id', 'ecommerce_transaction_id', 'ecommerce_refund_value_in_usd']],
    how='left',
    on=['user_pseudo_id', 'ecommerce_transaction_id']
)
df_merge_final.head()

user_pseudo_id ecommerce_transaction_id  \
0  10111055.8768683862                   741471   
1   1019527.5799124267                     2105   
2   1021887.1151788384                    76937   
3   1026932.0858862293                   339943   
4   1026932.0858862293                    29460   

   transaction_ga_session_number  user_ltv_revenue  \
0                          1.000            94.000   
1                          1.000            44.000   
2                          1.000            21.000   
3                          1.000           110.000   
4                          3.000           214.000   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      10.000                 94.000                0.000   
1                       5.000                 32.000                0.000   
2                       2.000                 10.000                1.000   
3                       4.000                110.000                2.000   
4                       4.000                 55.000                0.000   

   sum_item_quantity  sum_event_name_begin_checkout  \
0              3.000                          6.000   
1              7.000                          8.000   
2              2.000                          9.000   
3              2.000                          2.000   
4              1.000                          4.000   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      12.000                       1.000   
1                       0.000                       1.000   
2                       0.000                       0.000   
3                       0.000                       0.000   
4                      27.000                       0.000   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                                0.000                                0.000   
1                                0.000                                0.000   
2                                0.000                                0.000   
3                                1.000                                0.000   
4                                1.000                                0.000   

   sum_device_mobile_brand_name_Microsoft  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3                                   0.000   
4                                   0.000   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                 0.000                                0.000   
1                                 1.000                                0.000   
2                                 0.000                                0.000   
3                                 0.000                                0.000   
4                                 0.000                                0.000   

   sum_device_mobile_model_name_ChromeBook  \
0                                    0.000   
1                                    0.000   
2                                    0.000   
3                                    0.000   
4                                    0.000   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                                0.000                              0.000   
1                                0.000                              0.000   
2                                1.000                              0.000   
3                                0.000                              0.000   
4                                0.000                              0.000   

   sum_device_mobile_model_name_iPhone  \
0                                0.000   
1                                0.000   
2                                0.000   
3                                0.000   
4                                0.000

In [ ]:
# create table schema to prevent errors when uploading data
schema = []

for col, dtype in zip(df_merge_final.dtypes.index, df_merge_final.dtypes.values):
    col_dict = {}
    col_dict['name'] = col

    if dtype == object:
        typ = 'STRING'
    elif dtype == bool:
        typ = 'BOOLEAN'
    elif dtype == 'datetime64[ns]':
        typ = 'DATETIME'
    elif dtype in (int, 'Int64', 'uint8'):
        typ = 'INTEGER'
    elif dtype == float:
        typ = 'FLOAT'
    else:
        print(dtype)
    col_dict['type'] = typ

    schema.append(col_dict)

In [ ]:
# df_merge_final.to_gbq(f'{project_name}.return_prediction_ga4.step_5_merged_final',
#                         project_id=project_name,
#                         if_exists='replace',
#                         location=region,
#                         chunksize=100_000,
#                         table_schema=schema)